src/preact/base-element.js

```diff
@@ -617,6 +636,117 @@ export class PreactBaseElement extends AMP.BaseElement {
     }
     return this.defaultProps_[prop];
   }
+
+  /**
+   * Returns reference to upgraded imperative API object, as in React's
+   * useImperativeHandle.
+   * @return {!Promise<!API_TYPE>}
+   */
+  getApi() {
+    const api = this.apiWrapper_;
+    if (api) {
+      return Promise.resolve(api);
+    }
+    if (!this.deferredApi_) {
+      this.deferredApi_ = new Deferred();
+    }
+    return this.deferredApi_.promise;
+  }
+
+  /**
+   * Creates a wrapper around a Preact ref. The API surface exposed by this ref
+   * **must** be consistent accross all rerenders.
+   *
+   * This wrapper is necessary because every time React rerenders, it creates
+   * (depending on deps checking) a new imperative handle and sets that to
+   * `ref.current`. So if we ever returned `ref.current` directly, it could go
+   * stale by the time its actually used.
+   *
+   * @param {!API_TYPE} current
+   * @private
+   */
+  initApiWrapper_(current) {
+    const api = map();
+    const keys = Object.keys(current);
+    for (let i = 0; i < keys.length; i++) {
+      const key = keys[i];
+      wrapRefProperty(this, api, key);
+    }
+    this.apiWrapper_ = api;
+    if (this.deferredApi_) {
+      this.deferredApi_.resolve(api);
+      this.deferredApi_ = null;
+    }
+  }
+
+  /**
+   * Verifies that every Preact render exposes the same API surface as the previous render.
+   * If it does not, the API wrapper is syncrhonized.
+   *
+   * @param {!API_TYPE} current
+   * @private
+   */
+  checkApiWrapper_(current) {
+    if (!getMode().localDev) {
+      return;
+    }
+    const api = this.apiWrapper_;
+    const newKeys = Object.keys(current);
+    for (let i = 0; i < newKeys.length; i++) {
+      const key = newKeys[i];
+      devAssert(
+        hasOwn(api, key),
+        'Inconsistent Bento API shape: imperative API gained a "%s" key for %s',
+        key,
+        this.element
+      );
+    }
+    const oldKeys = Object.keys(api);
+    for (let i = 0; i < oldKeys.length; i++) {
+      const key = oldKeys[i];
+      devAssert(
+        hasOwn(current, key),
+        'Inconsistent Bento API shape: imperative API lost a "%s" key for %s',
+        key,
+        this.element
+      );
+    }
+  }
+}
+
+/**
+ * @param {tyepof PreactBaseElement} baseElement
+ * @param {!Object} api
+ * @param {string} key
+ */
+function wrapRefProperty(baseElement, api, key) {
+  Object.defineProperty(api, key, {
+    configurable: true,
+
+    get() {
+      return baseElement.currentRef_[key];
+    },
+
+    set(v) {
+      baseElement.currentRef_[key] = v;
+    },
+  });
+}
+
+/**
+ * Returns the upgraded imperative API object, once Preact has actually mounted.
+ *
+ * This technically works with both Bento and Legacy components, returning the
+ * BaseElement instance in the later case.
+ *
+ * @param {!Element} el
+ * @return {!Promise<!Object>}
+ */
+export function whenUpgraded(el) {
```

Comment:
This looks like something that belongs to the `src/base-element.js` since it's fully universal? We can name it `whenApiReady` to be more precise. However, a bigger question I have: how do we make it available to external user? Initially, at least, I'm ok if we add `getApi` to the `CustomElement` and instruct users to call `whenDefined()` themselves. But I'd like to get a better idea of the plan for this.

Reply:
We had discussed attaching this to a global `Bento` object, but that's not necessary in AMP mode. So registering it would fall onto the auto-envolope format to provide the API globally. I can add it here, but it wouldn't be necessary.

Reply:
Sure, but with dev-preview based on v0.js we need some way to access the `getApi()`. So I see a few of options:

1. Expose it on `CustomElement` and advise users to do `ce.whenDefined(e.tagName).then(() => e.getApi())`
2. Expose it on `AMP` global. Though, unfortunately, it can only be used via `AMP.push(() => {...})`.
3. Go ahead and publish a npm package with just this method.

I think given the future plans, the option (1) is a good compromise.

Reply:
Sure, I'll add it now.